### PR TITLE
reduce memory fragmentation 

### PR DESCRIPTION
**Pull request by CLFutureX**
Motivation:
  
1. Currently, when allocating from a magazine, if the remaining memory of the current chunk is not enough for allocation,
   the chunk is directly released.
   If the size to be allocated at this time is greater than the remaining memory of the chunk but >= RETIRE_CAPACITY, direct release will result in significant memory fragmentation.

2. When transferring the current chunk, if the transfer to the central queue ultimately fails, the chunk is chosen to be released.

Modification:
1. In this scenario, consider transferring the chunk to the nextInLine.

2. When transferring the chunk, if the transfer to the central queue still fails, consider comparing it with the local nextInLine and retiring the smaller one.

### CODE REVIEW

**Comment by franz1981** [comment]
I like the idea of not wasting it - well done!

**Comment by chrisvest** [comment]
The "trans result" int and interface is unnecessary, because we only make a single boolean decision based on it.
Please try to simplify the change before I look at it further, as I think this is more complicated than it needs to be.

**Comment by CLFutureX** [comment]
> The "trans result" int and interface is unnecessary, because we only make a single boolean decision based on it.
> Please try to simplify the change before I look at it further, as I think this is more complicated than it needs to be.

Well, to avoid picking up a chunk that was just put into nextInLine or the central queue, the TRAN_RESULT class was added at the time. Now, by readjusting the order to get before putting, the issue can be resolved.

File: buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java [hunk 1]
```diff
@@ -572,6 +572,23 @@ private void allocate(int size, int sizeBucket, int maxCapacity, AdaptiveByteBuf
             }
         }
 
+        private void transChunk(Chunk current) {
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:575]:
```suggestion
        private void transferChunk(Chunk current) {
```

File: buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java [hunk 2]
```diff
@@ -554,12 +558,8 @@ private void allocate(int size, int sizeBucket, int maxCapacity, AdaptiveByteBuf
...
                         curr.release();
                         current = newChunk;
-                    } else if (!trySetNextInLine(newChunk)) {
-                        if (!parent.offerToQueue(newChunk)) {
-                            // Next-in-line is occupied AND the central queue is full.
-                            // Rare that we should get here, but we'll only do one allocation out of this chunk, then.
-                            newChunk.release();
-                        }
+                    } else {
+                        transChunk(newChunk);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:562]:
```suggestion
                        transferChunk(newChunk);
```

File: buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java [hunk 3]
```diff
@@ -533,9 +532,14 @@ private void allocate(int size, int sizeBucket, int maxCapacity, AdaptiveByteBuf
...
             }
             current = curr;
-
             assert current != null;
-
+            if (last != null) {
+                if (last.remainingCapacity() < RETIRE_CAPACITY) {
+                    last.release();
+                } else {
+                    transChunk(last);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:540]:
```suggestion
                    transferChunk(last);
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
I like the idea... Just some nits related to naming the method

**Comment by CLFutureX** [comment]
> I like the idea... Just some nits related to naming the method

Thanks for your suggestion，its great

**Review by normanmaurer — APPROVED** [review]
LGTM

**Comment by normanmaurer** [comment]
@chrisvest PTAL

File: buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java [hunk 4]
```diff
@@ -572,6 +572,23 @@ private void allocate(int size, int sizeBucket, int maxCapacity, AdaptiveByteBuf
             }
         }
 
+        private void transferChunk(Chunk current) {
+            if (NEXT_IN_LINE.compareAndSet(this, null, current)
+                    || parent.offerToQueue(current)) {
+                return;
+            }
+            Chunk nextChunk = NEXT_IN_LINE.get(this);
+            if (current.remainingCapacity() > nextChunk.remainingCapacity()) {
```
[the remark this answers is not in the document]
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:581]: yess

[the remark this answers is not in the document]
Comment (chrisvest) [reply buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:581]: The magazine should have exclusive access to the nextInLine chunks, so this should not be a problem to access the chunk. However, with #14334 I think it'll be possible for the `NEXT_IN_LINE.compareAndSet` to fail and for the `NEXT_IN_LINE.get` to then return `null`, so it'd be good to have a null-check here.

[the remark this answers is not in the document]
Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:581]: Good point ... @CLFutureX please add a null check as @chrisvest suggested

[the remark this answers is not in the document]
Comment (CLFutureX) [reply buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java]: Good suggestion, I've added it

**Review by normanmaurer — APPROVED** [review]
LGTM

**Comment by normanmaurer** [comment]
@CLFutureX thanks a lot!

**Comment by normanmaurer** [comment]
I think we also want to port this to main... @chrisvest WDYT ?

**Comment by chrisvest** [comment]
@normanmaurer I agree.

**Comment by normanmaurer** [comment]
@CLFutureX you want to do a PR against main ?

**Comment by CLFutureX** [comment]
@normanmaurer  with pleasure,i will do it later

**Pull request merged**